src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -725,8 +750,31 @@ where
 
                 let mut batch_output = batch_output.activate();
                 let mut summary_output = summary_output.activate();
+                let mut health_output = health_output.activate();
 
                 for (message_batch, source_upper) in buffer.drain(..) {
+                    if let Some(batch) = &*message_batch.borrow() {
+                        let has_errors = batch.non_definite_errors.first();
+                        let has_messages = batch.messages.values().any(|vs| !vs.is_empty());
+
+                        let maybe_health = match (has_errors, has_messages) {
+                            (Some(error), _) => {
+                                // Arguably this case should be "failed", since generally a source
+                                // cannot recover by the time an error reaches this far down the pipe.
+                                // However, we don't actually shut down the source on error yet, so
+                                // treating this as a possibly-temporary stall for now.
+                                Some(HealthStatus::StalledWithError(error.error.to_string()))
+                            }
+                            (_, true) => Some(HealthStatus::Running),
+                            (None, false) => None,
+                        };
+
+                        if let Some(health) = maybe_health {
+                            let mut session = health_output.session(&cap);
```

Comment:
It might be a bit faster if you open this session at the beginning of the the operator, but this might not be measurable.

Also, this is subtlely wrong, but in a way that sometimes does not manifest; it needs to be a capability for the correct output port, like the
```
let summary_cap = cap.delayed_for_output(cap.time(), summary_output_port);
```
below!

Reply:
> It might be a bit faster if you open this session at the beginning of the the operator, but this might not be measurable.

Interesting! You mean lifting this to the beginning of the `for_each` block, where the outputs are activated? (And would the other sessions we open below be able to do the same thing?)

> Also, this is subtlely wrong, but in a way that sometimes does not manifest

Yikes! Will fix. Is there a brief summary of _why_ this is incorrect, and how it might go wrong?

Reply:
@bkirwi its some cursed timely details that ill admit I dont understand fully!